webauthn/credential.go

```diff
@@ -48,6 +61,9 @@ type CredentialFlags struct {
 	// Flag BS indicates the credential has been backed up and/or sync'd. This value can change but it's recommended
 	// that RP's keep track of this value.
 	BackupState bool `json:"backupState"`
+
+	// Raw flags which can be used to repopulate the flags.
+	Raw protocol.AuthenticatorFlags `json:"raw"`
```

Comment:
Thank you for this improvements.

After reading Gitea's code again, I managed to store/restore the CredentialFlags directly by JSON, without the Raw field.

Is it safe&good enough? If it is ok, then maybe we do not need to introduce this Raw fields, it seems make the logic more complex.

However, if "Raw" is a must for compatible reasons, then please add it as your plan

(Indeed I am not quite famliar with webauthn's code and design)

Thank you again.

Reply:
I think that it is good enough in most instances. The only advantage here is that if the W3C includes other flags from the current list in ones that must be stored then you may lose those. We only include the ones in that struct that are on the credential record list or have been discussed in the working group as possibly being added to that list.

Technically also allows efficient storage of the flags, which may be preferable to some users. What I might do is exclude the raw value from the JSON fields by removing the tag and making it private and in that way it's optional for users to restore it in this way.

Reply:
Oh yes, now I could understand more and I think it is useful, because I just copied it to https://github.com/go-gitea/gitea/pull/32623 😁